### DISCUSSION TITLE
sd_bus_attach_event wrapper

### DISCUSSION THREAD

**Discussion by hadrianw** [Kistler-Group/sdbus-cpp#262] [Ideas]
Hello,

First I would like you thank you for this project.

I would like to ask if you have considered an sd_bus_attach_event wrapper? At my employer we use your nice project and we started to migrate also to sd-event for our event loops. For our needs I made a wrapper for it. It is not yet decided whether or not we will go with it.

I was wondering if you would consider accepting a pull request with it. Also if yes what would be required by you in terms of the implementation. For now I have a non-hermetic API for it: `void IConnection::attachEventLoop(sd_event *ev, int priority)`. I understand it does not really fit with the rest of the library, but if there are ways you would see it better I'm all ears.

Kind regards and thank you,
Hadrian Węgrzynowski

**Comment by dleeds-cpi** [comment]
Personally, I think this addition would provide value and opportunities to optimize. On my current project, I have created a C++ wrapper around `sd-event` and a wrapper around `IConnection` that provides the event loop integration. This is rather convoluted, and it would be a much simpler option to attach directly to the underlying event loop.

To avoid introducing dependencies to the `systemd/sd-event.h` header in `sdbus-c++`, I would forward declare the `sd_event` struct, like this:

```c++
struct sd_event;
void IConnection::attachEventLoop(struct sd_event *e, int priority);
```

**Comment by sangelovic** [comment]
@hadrianw Thank you for kind words. Would you like to prepare a PR for this along the lines of @dleeds-cpi 's comment?

**Reply by hadrianw** [reply]
I will submit a pull request today or tomorrow.

**Reply by hadrianw** [reply]
Sorry for the delay. Here is my pull request. I didn't manage to add tests for it, as I had trouble setting up the environment for them to work.

https://github.com/Kistler-Group/sdbus-cpp/pull/271
